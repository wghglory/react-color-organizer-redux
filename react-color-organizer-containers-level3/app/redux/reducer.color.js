import C from './constants.color';

export const colorReducer = (state = {}, action) => {
  switch (action.type) {
    case C.ADD_COLOR:
      return {
        id: action.id,
        title: action.title,
        color: action.color,
        timestamp: action.timestamp,
        rating: 0
      };
    case C.RATE_COLOR:
      return (state.id !== action.id) ?
        state :
        // Object.assign({}, state, { rating: action.rating });
        {
          ...state,
          rating: action.rating
        };
    default:
      return state;
  }
};

export const colorsReducer = (state = [], action) => {
  switch (action.type) {
    case C.ADD_COLOR:
      return [
        ...state,
        colorReducer({}, action)
      ];
    case C.RATE_COLOR:
      return state.map(
        c => colorReducer(c, action)
      );
    case C.REMOVE_COLOR:
      return state.filter(
        c => c.id !== action.id
      );
    default:
      return state;
  }
};