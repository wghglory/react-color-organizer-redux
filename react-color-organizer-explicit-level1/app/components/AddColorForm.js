import React from 'react';
import PropTypes from 'prop-types';
import '../scss/AddColorForm.scss';

// use store, need action
import { addColor } from '../redux/actions.color';

const AddColorForm = ({ store }) => {
    let _title, _color;

    const submit = e => {
        e.preventDefault();
        store.dispatch(addColor(_title.value, _color.value)); // use store
        _title.value = '';
        _color.value = '#000000';
        _title.focus();
    };

    return (
        <form className="add-color" onSubmit={submit}>
            <input ref={input => _title = input}
                type="text"
                placeholder="color title..." required />
            <input ref={input => _color = input}
                type="color" required />
            <button>ADD</button>
        </form>
    );

};

AddColorForm.propTypes = {
    store: PropTypes.object
};

export default AddColorForm;