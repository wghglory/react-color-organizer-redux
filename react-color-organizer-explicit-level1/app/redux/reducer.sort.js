import C from './constants.color';

export const sortReducer = (state = "SORTED_BY_DATE", action) => {
  switch (action.type) {
    case C.SORT_COLORS:
      return action.sortBy;
    default:
      return state;
  }
};

