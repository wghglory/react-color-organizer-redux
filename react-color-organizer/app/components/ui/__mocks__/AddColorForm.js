import React from 'react';
const AddColorFormMock = () => <div></div>;
AddColorFormMock.displayName = "AddColorFormMock";
export default AddColorFormMock;