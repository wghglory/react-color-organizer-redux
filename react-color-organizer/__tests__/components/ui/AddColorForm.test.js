import { shallow, mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import { compose } from 'redux';
import AddColorForm from '../../../app/components/ui/AddColorForm';

describe("<AddColorForm /> UI Componnet", () => {

  let wrapper;

  describe("Rendering the UI", () => {

    it("renders correctly", () => {
      compose(expect, toJSON, shallow)(<AddColorForm />)
        .toMatchSnapshot();
    });

    it("Submitting default onNewColor does not cause error", () => {
      wrapper = mount(<AddColorForm />);
      wrapper.find('input').get(0).value = 'test-color';
      wrapper.find('input').get(1).value = '#0000FF';
      // wrapper.ref('_title').get(0).value = 'test-color'
      // wrapper.ref('_color').get(0).value = '#0000FF'
      wrapper.find('form').simulate('submit');
    });

  });

  describe("Adding New Colors", () => {

    const _addColor = jest.fn();

    beforeAll(() => {
      wrapper = mount(<AddColorForm onNewColor={_addColor} />);
      // wrapper.ref('_title').get(0).value = 'test-color'
      // wrapper.ref('_color').get(0).value = '#000099'
      wrapper.find('input').get(0).value = 'test-color';
      wrapper.find('input').get(1).value = '#000099';
      wrapper.find('form').simulate('submit');
    });

    it("adds colors correctly", () =>
      expect(_addColor).toBeCalledWith('test-color', '#000099')
    );

    it("resets the title value", () =>
      // expect(wrapper.ref('_title').get(0).value).toBe("")
      expect(wrapper.find('input').get(0).value).toBe("")
    );

    it("resets the color value", () =>
      // expect(wrapper.ref('_color').get(0).value).toBe("#000000")
      expect(wrapper.find('input').get(1).value).toBe("#000000")
    );

  });

});
