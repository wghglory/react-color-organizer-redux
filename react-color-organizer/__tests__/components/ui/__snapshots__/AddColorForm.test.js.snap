// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddColorForm /> UI Componnet Rendering the UI renders correctly 1`] = `
<form
  className="add-color"
  onSubmit={[Function]}
>
  <input
    placeholder="color title..."
    required={true}
    type="text"
  />
  <input
    required={true}
    type="color"
  />
  <button>
    ADD
  </button>
</form>
`;
